[채팅면접] 음성 채팅 모드 전환 및 녹음/취소 버튼 기능 추가

yundal8755 opened the pull request:
## 📝 변경 내용
- https://github.com/MakeFrog/TechTalk/issues/118

### UI 큰 틀 구현
- 타이핑 모드에서 마이크 버튼을 누르면 스피치 모드로 전환되도록 구현
- 스피치 모드의 UI 구현 (메인 녹음 버튼, 타이핑 모드 버튼, 녹음 취소/정지 버튼)
- Speech_to_Text의 결과물를 화면에 표시
- 타이핑 모드 ToolTip 표시 (초안)

### Speech_to_Text 기능 구현
- 스피치 모드 진입 시 음성 녹음 및 음성 인식 권한 요청
- 메인 녹음 버튼 클릭 시 터미널에서 실시간 음성 인식 결과를 표시하고, 메인 녹음 버튼을 다시 누르면 녹음을 정지하고 화면에 결과를 출력

### 녹음 취소/정지 버튼 구현하기
- isListening이 false일 때 ⇒ 버튼 비활성화
- isListening이 true일 때 ⇒ 버튼 활성화 ⇒ 취소 메서드 구현

<br>

## 🔍 시연 영상
https://github.com/user-attachments/assets/79383e1c-d7b8-4842-9ba0-fb054203654a

<br>

## 👥 리뷰어
@Yellowtoast , @Xim-ya 

<br>

## 📌 기타 사항
- 테스트시 반드시 마이크 사용을 할 수 있는 기기(맥미니x 노트북o)에서 사용해주세요. 
- 마이크가 없는 경우 speech to text 기능이 작동하지 않으며, 이에 대한 예외 처리 로직은 아직 구현되지 않았습니다.
- 현재 ChatState와 ChatEvent로 분리되어야 할 로직이 UI에 남아있는데, 이는 추후 리팩토링할 예정입니다.
- 타이핑 모드 ToolTip의 위치가 시뮬레이터마다 다르게 표시되는데, 해당 문제 또한 추후에 해결할 예정입니다.

<br>

yundal8755 committed:
feature : Speech-to-Text 기능 추가 및 음성 모드 UI 구현

- speech_to_text 패키지 추가
- iOS 마이크 접근 허용권한 설정
- [onMicBtnTapped] : 마이크 버튼 클릭시 이벤트 추가
- [InterviewTabView] : 마이크 버튼 클릭시 음성 모드 UI 추가

yundal8755 committed:
feature : UI 에셋 추가 및 녹음 취소 버튼 기능 구현

- UI 에셋 추가 : 타이핑 모드 Aa, toolTip, 녹음 취소 svg
- isListening bool값에 따른 녹음 취소 버튼 활성화/비활성화 기능 구현

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
## P1 소스파일 분리

한 소스파일 안에 너무 많은 로컬 위젯들이 위치하고 있어 가독성이 좋지 않아 보입니다 (500줄 초과)
각 로컬 위젯을 별도의 소스파일로 분리해주시면 좋을 것 같아요.

- widgets
  - interview_tab_view
     - interview_tab_View.dart
     - bottom_speech_to_text_field.dart
     - bottom_input_field.dart

위 폴더 구조를 적용하는건 어떤지 검토해주세요~

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -94,17 +113,21 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+        data: (_) => _buildTextField(progressState(ref), isSpeechMode),
+        error: (_, __) =>
+            _buildTextField(InterviewProgress.error, isSpeechMode),
+        loading: () => _buildTextField(InterviewProgress.initial, isSpeechMode),
```
## P2 불필요한 ValueNotifier 상태 전달 + 음성 인식 활성화 조건별 예외처리 로직 추가 필요

로딩중이거나 예상하지 못한 오류가 발생할 경우 음성 인식 기능이 작동하면 안됩니다.

```dart
     child: chatAsyncAdapterValue(ref).when(
        data: (_) => _buildTextField(progressState(ref), isSpeechMode),
        error: (_, __) =>
            _buildTextField(InterviewProgress.error),
        loading: () => _buildTextField(InterviewProgress.initial),
      ),
    );
  }

  ///
  /// 입력폼
  ///
  Widget _buildTextField(
    InterviewProgress progressState,
    [ValueNotifier<bool>? isSpeechMode] // <-- CHECK!
  ) {
```

위 코드처럼 오류가 발생하거나 로딩중일 경우에는
speechMode로 전환 여부를 결정하는 상태인 `isSpeechMode` 값을 전달하지 않는 방식을 고려해보시면 좋을 것 같아요.

> isSpeechMode 파라미터를 nullable하게 설정하고 만약 null하다면 오류나 로딩중인것으로 판단하고 speechMode 전환을 막야합니다!

Xim-ya replied:
## P4 ref를 넘겨주지 않고 Consumer 사용
```dart
     ),
      child: chatAsyncAdapterValue(ref).when(
        data: (_) => _buildTextField(progressState(ref), ref),
        error: (_, __) => _buildTextField(InterviewProgress.error, null),
        loading: () => _buildTextField(InterviewProgress.initial, null),
      ),
    );
  }

  ///
  /// 입력폼
  ///
  Widget _buildTextField(InterviewProgress progressState, WidgetRef? ref) {
    return HookBuilder(
      builder: (context) {
```

이렇게 ref를 파라미터를 nullable하게 설정해서 조건별로 넘겨주어도 되지만, 
ref를 넘겨주지 않고 ref를 참조하는 위젯을 Consumer로 감싸는 방식이 조금 더 적합해보입니다!

```dart
Widget _buildTextField(InterviewProgress progressState) {
    return HookBuilder(
      builder: (context) {
        /// 하이라이트(마이크 기능 알림) 효과 노출 여부
        final showHighlightEffect = useState(isFirstInterview());
        final messageController = useTextEditingController();
        final focusNode = useFocusNode();

         ....
      
        return Stack(
          clipBehavior: Clip.none,
          alignment: Alignment.bottomRight,
          children: [
          .....
            /// MICROPHONE BUTTON
            AnimatedPositioned(
              left: isFieldFocused ? -50 : 0,
              duration: expandInOutDuration,
              child: AnimatedOpacity( 
                duration: const Duration(milliseconds: 120),
                opacity: isFieldFocused
                // || isSpeechMode(ref
                 ? 0 : 1,
                child: Consumer(
                  builder: (context, ref , _) {                // <- CHECK !!!
                    return ShrinkGestureView( 
                      onTap: () {
                       ref!.read(isSpeechModeProvider.notifier).toggle();
                      },
                      borderRadius: BorderRadius.circular(22),
                      child: Container(
                        height: 44,
                        width: 44,
                        decoration: BoxDecoration(
                          color: AppColor.of.blue1,
                          shape: BoxShape.circle,
                        ),
                        child: Stack(
                          children: [
                            if (showHighlightEffect.value.isTrue)
                              const GradientShineEffectView(),
                            Positioned(
                              child: Center(
                                child: SvgPicture.asset(
                                  Assets.iconsIconMic,
                                  colorFilter: ColorFilter.mode(
                                    showHighlightEffect.value
                                        ? AppColor.of.white
                                        : AppColor.of.brand3,
                                    BlendMode.srcIn,
                                  ),
                                ),
                              ),
                            ),
                          ],
                        ),
                      ),
                    );
                  },
                ),
              ),
            ),

            ....
              ),
            ),
          ],
        );
      },
    );
  }
```

이렇게하면 ref를 넘겨주지 않아도 되어서 코드가 조금 더 깔끔해지고, 
context를 분리해서 불필요한 rebuild도 줄일 수 잇을 것 같아요~

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -71,17 +82,25 @@ class InterviewTabView extends HookConsumerWidget with ChatState, ChatEvent {
+        if (isSpeechMode.value)
+          _BottomSpeechToTextField(isSpeechMode: isSpeechMode)
+        else
+          _BottomInputField(isSpeechMode: isSpeechMode),
```
## P2 음성 인식 활성 UI로 변경할 때 자연스러운 트랜지션 적용

음성 인식 활성화 모드로 들어갈 때 UI가 자연스럽게 나타나는 트랜지션을 적용해주시면 좋을 것 같아요!

이전에 `AnimatedSwitcher`같은 위젯을 사용하면 `_BottomSpeechToTextField`이 위젯이 활성화 되지 않았는데도 초기화되어 권한을 요청하는 팝업을 노출한다는 문제가 있다고 말씀해주셨는데요. 

```dart
      AnimatedCrossFade(
          firstChild:  _BottomInputField(isSpeechMode: isSpeechMode),
          secondChild:   isSpeechMode.value  ? _BottomSpeechToTextField(isSpeechMode: isSpeechMode) : const EmptyBox(), // <-- CHECK
          crossFadeState: isSpeechMode.value ? CrossFadeState.showSecond : CrossFadeState.showFirst,
          duration: const Duration(
            milliseconds: 400,
          ),
        ),
```

```dart
isSpeechMode.value  ? _BottomSpeechToTextField(isSpeechMode: isSpeechMode) : const EmptyBox(),
```
이렇게 speechMode가 활성화 안되어 있다면 빈 위젯을 리턴하도록 하면 해당 문제를 해결할 수 있을 것 같습니다~

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -261,3 +286,231 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+  Widget build(BuildContext context, WidgetRef ref) {
+    // SpeechToText 객체 생성
+    final speechToText = useMemoized(stt.SpeechToText.new, []);
+    final isListening = useState(false);
```
## P3 상태 조작 위치 변경

현재 `isListening` 이라는 상태를 useState으로 선언해서 UI를 분기처리 해주고 계시는데요.
 isListening 변수 상태를 조작하는 위치를 변경하면 코드가 조금 더 깔끔해지고 안정성이 높아질 것 같아요.
```dart
      speechToText.stop();
      isListening.value = false;
```
지금 처럼 speechToText Controller도 조작하고 isListening 상태도 조작하는게 아니라

```dart
    bool available = await speechToText.initialize(
          onStatus: (status) {
            if (status == 'listening') {
              isListening.value = true;
            }
            if (status == 'done') {
              isListening.value = false;
            }
            print('Speech status: ==> $status');
          },
          onError: (error) => print('Speech error: $error'),
        );
```

speechToText를 init해주는 부분에서 isListening 상태를 조작하는 로직을 넣어주면 코드가 훨씬 코드가 깔끔해질 것 같습니다.
그리고 speechToText Controller의 상태를 조금 더 안전하게 싱크를 맞출 수 있을 것 같고요.

한번 고민해주세요 :)

yundal8755 replied:
- 아래 내용처럼 startListening()와 stopListening()의 메서드에는 isListening 상태를 따로 조작하지 않도록 수정했습니다.

```dart
    // SpeechToText 초기화 및 리스닝 상태 처리
    useEffect(() {
      Future<void> initSpeech() async {
        bool available = await speechToText.initialize(
          onStatus: (status) {
            if (status == 'listening') {
              isListening.value = true;
            }
            if (status == 'done') {
              isListening.value = false;
            }
            print('Speech status: $status');
          },
          onError: (error) => print('Speech error: $error'),
        );

        if (available) {
          print('초기화됨');
        } else {
          print('사용 불가능');
        }
      }

      initSpeech();
      return null;
    });
```

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -261,3 +286,231 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+      child: Column(
+        children: [
+          if (!isListening.value && recognizedText.value.isNotEmpty)
+            _buildRecognizedText(recognizedText.value),
+          _buildSpeechField(
+            isSpeechMode,
+            showHighlightEffect,
+            isListening.value,
+            recognizedText.value,
+            startListening,
+            stopListening,
+          ),
```
## P3 메소드 argument 최소화
위젯을 리팩토링 해주신건 좋지만 
위젯을 리턴하는 `_buildSpeechField`  메소드에 너무 많은 argument가 있어서 가독성을 해치는 것 같습니다.
dart lint를 참고해서 말씀드리자면 5개 이상의 argument를 전달하는건 좋지 않는다고 합니다. 

전달하는 argument를 어떻게하면 최소화할 수 있을지 고민해보면 좋을 것 같아요.
> 개인적으로 _buildSpeechField 위젯을 굳이 메소드로 따로 빼지 않아도 될 것 같다는 생각이 듭니다.

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -261,3 +286,231 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    // SpeechToText 객체 생성
+    final speechToText = useMemoized(stt.SpeechToText.new, []);
```
## useMemoized 사용 이유
혹시 speechToText controller를 초기화 하실 때  useMemoized을 이용하신 이유가 있을까요?

yundal8755 replied:
useMemorized가 초기화를 한 번만 실행하는 것을 강조할 때 사용하는 걸로 알고 있었는데, 제가 이해하고 사용하는게 맞는지 궁금하네요. final speechToText = stt.SpeechToText(); 이 방식이 객체를 매번 새로 생성하는 걸로 알고 있고, 이게 성능에 조금이라도 영향을 줄까 싶어서 작성해보았습니다. 피드백 부탁드립니다.

Xim-ya replied:
오... 그렇네요.
저렇게 접근하면 빌드가 여러번 되어도 객체에 값을 여러번 할당하지 않겠군요.

하나 배워갑니다 :)

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -261,3 +286,231 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+
+    // SpeechToText 초기화 및 리스닝 상태 처리
+    useEffect(() {
+      Future<void> initSpeech() async {
+        bool available = await speechToText.initialize(
+          onStatus: (status) => print('Speech status: $status'),
+          onError: (error) => print('Speech error: $error'),
+        );
+
+        if (available) {
+          print('초기화됨');
+        } else {
+          print('사용 불가능');
+        }
+      }
+
+      initSpeech();
```
## P1 Exception에 따른 예외처리 로직  (Resolved 필요 X)

저번 회의 때 잠깐 말씀드렸지만 어떤 Exception이 떨어질 수 있는 확인하시고
그에 맞는 예외처리 로직을 적용해주시면 좋을 것 같습니다!

> 지금 PR에 해당 내용을 적용해주셔도 되고, 다음 PR 작업해주셔도 됩니다!

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -261,3 +286,231 @@ class _BottomInputField extends HookConsumerWidget with ChatState, ChatEvent {
+                  const Gap(48),
+
+                  // 녹음 취소
+                  GestureDetector(
+                    onTap: () {
+                      print('현재 SpeechMode인가? : ${isSpeechMode.value}');
+
+                      isListening
+                          ? stopListening()
+                          : print('녹음 취소 버튼이 현재 비활성화 되어있음');
+                    },
+                    child: CircleAvatar(
+                      radius: 24,
+                      backgroundColor: isListening
+                          ? AppColor.of.red1
+                          : AppColor.of.background1,
+                      child: SvgPicture.asset(
+                        Assets.iconsDeleteOrWrong,
+                        colorFilter: ColorFilter.mode(
+                          isListening ? AppColor.of.red2 : AppColor.of.gray3,
+                          BlendMode.srcIn,
+                        ),
+                        // color: AppColor.of.gray3,
```
## P4 자연스러운 트래진션 적용
AnimatedContainer을 이용해서 버튼의 배경색을 조금 더 자연스럽게 변경해주는 트랜지션을 적용하는건 어떤가요?

yundal8755 replied:
넵 AnimatedContainer 적용했습니다!

```dart
                        // 녹음 취소
                        GestureDetector(
                          onTap: () {
                            isListening.value
                                ? stopListening()
                                : print('녹음 취소 버튼이 현재 비활성화 되어있음');
                          },
                          child: AnimatedContainer(
                            duration: const Duration(milliseconds: 300),
                            decoration: BoxDecoration(
                              shape: BoxShape.circle,
                              color: isListening.value
                                  ? AppColor.of.red1
                                  : AppColor.of.background1,
                            ),
                            child: CircleAvatar(
                              radius: 24,
                              backgroundColor: Colors.transparent,
                              child: SvgPicture.asset(
                                Assets.iconsDeleteOrWrong,
                                colorFilter: ColorFilter.mode(
                                  isListening.value
                                      ? AppColor.of.red2
                                      : AppColor.of.gray3,
                                  BlendMode.srcIn,
                                ),
                              ),
                            ),
                          ),
                        ),
```

Xim-ya commented on lib/presentation/pages/interview/chat/widgets/interview_tab_view.dart:
```diff
@@ -21,6 +24,8 @@ class InterviewTabView extends HookConsumerWidget with ChatState, ChatEvent {
 
   @override
   Widget build(BuildContext context, WidgetRef ref) {
+    final isSpeechMode = useState(false);
```
## P4 Prop Drilling 
isSpeechMode 라는 ValueNotifier의 상태를 여러 하위 위젯에 연달아 넘겨주고 있는데, 해당 상태를 별도의 Provider로 빼서 관리하는건 어떤가요?

계속 인자를 넘겨주지 않아도 되어서 코드가 조금 깔끔해질 것 같습니다. 

근데 동시에 isSpeechToText 같은 경우 하단 BottomTextField 공간에서 사용되는 상태이기 때문이 별도의 상태로 안빼도 될 것 같다는 생각이 들기도 하네요. 한번 고민해주세요 :)

yundal8755 replied:
_buildTextField()의 MICROPHONE BUTTON과 BottomSpeechToTextField의 타이핑 모드 이렇게 두 군데에서 사용되어 Provider로 빼서 관리하는 것으로 결정했습니다!

Xim-ya reviewed (approved):
수고하셨습니다!

yundal8755 committed:
fix : 모든 PR 코멘트 해결 및 코드 수정 완료

- 소스파일 분리
- 불필요한 ValueNotifier 상태 전달
- 음성 인식 활성 UI로 변경할 때 자연스러운 트랜지션 적용
- 상태 조작 위치 변경
- 메소드 argument 최소화
- useMemoized 사용 이유 설명
- 자연스러운 트래진션 적용

yundal8755 committed:
feature : 마이크 및 음성 인식 권한 핸들러 추가

yundal8755 committed:
fix : ref 대신 consumer 사용
